### PR TITLE
[JIT] Allow `tuple` and `list` generics

### LINKED ISSUE #98521: Unsupported annotation when using generics for typing
### 🐛 Describe the bug

When using [generics for typing](https://peps.python.org/pep-0585/#reference-implementation) in Python 3.9, I run into this error
```
Traceback (most recent call last):
  File "/Users/scripts/torchscript_bug.py", line 24, in <module>
    out = torch.jit.script(pad)
  File "/usr/local/Caskroom/miniforge/base/envs/ludwig/lib/python3.9/site-packages/torch/jit/_script.py", line 1341, in script
    fn = torch._C._jit_script_compile(
  File "/usr/local/Caskroom/miniforge/base/envs/ludwig/lib/python3.9/site-packages/torch/jit/annotations.py", line 366, in try_ann_to_type
    assert maybe_type, msg.format(repr(ann), repr(maybe_type))
AssertionError: Unsupported annotation typing.Union[int, tuple[int, int]] could not be resolved because None could not be resolved.
```
repro script
```
import torch
import torchvision.transforms.functional as F
from typing import Union


def to_tuple(v: Union[int, tuple[int, int]]) -> tuple[int, int]:
    """Converts int or tuple to tuple of ints."""
    if torch.jit.isinstance(v, int):
        return v, v
    else:
        return v


def pad(img: torch.Tensor, new_size: Union[int, tuple[int, int]]) -> torch.Tensor:
    new_size = to_tuple(new_size)
    old_size = img.shape[-2:]
    pad_size = (torch.tensor(new_size) - torch.tensor(old_size)) / 2
    padding = torch.cat((torch.floor(pad_size), torch.ceil(pad_size)))
    padding[padding < 0] = 0
    padding = [int(x) for x in padding]
    return F.pad(img, padding=padding, padding_mode="edge")


out = torch.jit.script(pad)
```

### Versions

`Python 3.9.16`
```
torch                                    2.0.0
torchaudio                               2.0.1
torchdata                                0.6.0
torchinfo                                1.7.2
torchmetrics                             0.11.4
torchtext                                0.15.1
torchvision                              0.15.1
```

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### PR DESCRIPTION
**Pull request by malfet**
As in Python-3.9+ `Dict`, `List`, and `Tuple` from `typing` module are deprecated in favor of their `builtins` counterparts, see [PEP 585](https://peps.python.org/pep-0585/)

Test plan: Run:
```
import torch
from typing import Union

@torch.jit.script
def to_tuple(v: Union[int, tuple[int, int]]) -> tuple[int, int]:
    """Converts int or tuple to tuple of ints."""
    if torch.jit.isinstance(v, int):
        return v, v
    else:
        return v

print(to_tuple(1), to_tuple((3, 4)))
```

It's almost impossible to add test to an existing CI, as test script will not be parseable by Python-3.8, which is a oldest supported Python version

Fixes https://github.com/pytorch/pytorch/issues/98521

### CODE REVIEW

**Review by kit1980 — APPROVED** [review]

File: torch/_jit_internal.py [hunk 1]
```diff
@@ -983,9 +985,12 @@ def is_tuple(ann) -> bool:
     # For some reason Python 3.7 violates the Type[A, B].__origin__ == Type rule
     if not hasattr(ann, "__module__"):
         return False
+
+    ann_origin = getattr(ann, "__origin__", None)
+    if IS_PY39_PLUS and ann.__module__ == "builtins" and ann_origin is tuple:
+        return True
     return ann.__module__ == "typing" and (
```
Comment (kit1980) [inline torch/_jit_internal.py:992]: Remove the parenthesis?

Comment (malfet) [reply torch/_jit_internal.py:992]: I need parenthesis for `or` to take precedence, otherwise linter suggests the same ;)

**Pull request closed without merging**